lib-extra/src/main/java/com/diffplug/spotless/extra/config/MavenCoordinates.java

```diff
@@ -0,0 +1,223 @@
+		private static final long serialVersionUID = 1L;
+
+		// Format of a dependency version restriction (type is always JAR)
+		public static final String FORMAT = "<groupId>:<artifactId>[:packaging][:classifier]:<versionRestriction>";
+		private final String groupId;
+		private final String artifactId;
+		private final String packagingOrNull;
+		private final String classifierOrNull;
+		private final String versionRange;
+		private String dependency = null;
+
+		/**
+		 * Converts the format {@code <groupId>:<artifactId>[:packaging][:classifier]} dependency and
+		 * {@code <versionRestriction>} value
+		 */
+		Coordinate(String dependency, String version) {
+			this(String.format("%s:%s", dependency, version));
+		}
+
+		/** Creates version restriction from string. String format is {@value #FORMAT}. */
+		Coordinate(String coordinate) {
+			if (null == coordinate) {
+				throw new UserArgumentException(null, "Maven coordinate");
+			}
+			LinkedList<String> dependencyParts = new LinkedList<String>(
+					Arrays.asList(coordinate.split(":", -1)));
+			if (dependencyParts.size() < 3 || dependencyParts.size() > 5) {
+				throw new UserArgumentException(dependencyParts,
+						String.format("Maven coordinate format '%s', does not allow %d elements.", FORMAT, dependencyParts.size()));
+			}
+			versionRange = trimAndCheckNotEmpty(coordinate, dependencyParts.pollLast());
+			groupId = trimAndCheckNotEmpty(coordinate, dependencyParts.pollFirst());
+			artifactId = trimAndCheckNotEmpty(coordinate, dependencyParts.pollFirst());
+			packagingOrNull = dependencyParts.isEmpty() ? null : trimAndCheckNotEmpty(coordinate, dependencyParts.pollFirst());
+			classifierOrNull = dependencyParts.isEmpty() ? null : trimAndCheckNotEmpty(coordinate, dependencyParts.pollFirst());
+		}
+
+		private static String trimAndCheckNotEmpty(String coordinate, String item) {
+			item = item.trim();
+			if (item.isEmpty()) {
+				throw new UserArgumentException(coordinate, String.format("Empty elements are not allowed by format '%s'.", FORMAT));
+			}
+			return item;
+		}
+
+		/** Returns Maven version (range) */
+		String getVersionRange() {
+			return versionRange;
+		}
+
+		/**
+		 * Get dependency string in the format
+		 * {@code <groupId>:<artifactId>[:packaging][:classifier]}
+		 */
+		String getDependency() {
+			if (null == dependency) {
+				StringBuilder dependencyVersionRestriction = new StringBuilder();
+				dependencyVersionRestriction.append(groupId).append(':').append(artifactId);
+				if (null != packagingOrNull) {
+					dependencyVersionRestriction.append(':').append(packagingOrNull);
+				}
+				if (null != classifierOrNull) {
+					dependencyVersionRestriction.append(':').append(classifierOrNull);
+				}
+				dependency = dependencyVersionRestriction.toString();
+			}
+			return dependency;
+		}
+
+		@Override
+		public String toString() {
+			return String.format("%s:%s", getDependency(), getVersionRange());
+		}
+
```

Comment:
Same up-to-date problem as earlier.

Reply:
I think you wanted to refer to the `hashCode`. The `toString` is just for readable exception messages.

Reply:
Correct!  I expected it to highlight above my click rather than below, apologies :)